packages/report-server/src/reportBuilder/output/functions/matrix/matrixBuilder.ts

```diff
@@ -0,0 +1,72 @@
+import { Matrix, Row } from '../../../types';
+import { MatrixParams } from './types';
+
+export class MatrixBuilder {
+  rows: Row[];
+
+  matrixData: Matrix;
+
+  params: MatrixParams;
+
+  constructor(rows: Row[], params: MatrixParams) {
+    this.rows = rows;
+    this.params = params;
+    this.matrixData = { columns: [], rows: [] };
+  }
+
+  assertIsString(values: unknown[]): asserts values is string[] {
+    values.forEach(value => {
+      if (typeof value !== 'string') {
+        throw new Error(`Expect value to be string, but got ${value}`);
+      }
+    });
+  }
+
+  buildBaseColumns() {
+    const assignColumnSetToMatrixData = (columns: unknown[]) => {
+      this.assertIsString(columns);
+      this.matrixData.columns = columns.map(c => ({ key: c, title: c }));
+    };
+
+    const { prefixColumns, nonColumnKeys } = this.params.columns;
+    if (Array.isArray(prefixColumns)) {
+      assignColumnSetToMatrixData(prefixColumns);
+      return this;
+    }
+    // When prefixColumns === '*', get columns from rows data
+    const columns = new Set();
```

Comment:
```suggestion
    const columns = new Set<string>();
```
^Might need to double check me on that sytax, but I think this will remove the need for you to do a string assertion on the columns later on 👍

Reply:
I've tried to but it seems like `Array.from(set)` is not smart enough to send back results with a string type.

Reply:
Thanks @rohan-bes , you are right! We can get rid of `assertIsString` function. 
I guess TypeScript sometimes is lagging to map the type, but totally that's my fault.